### PR TITLE
Create TIF Source Config API

### PR DESCRIPTION
**Pull request by jowg-amazon**
### Description
Preliminary architecture and flow for creating a job index and indexing a source config object.
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java [hunk 1]
```diff
@@ -103,9 +111,11 @@ public class SecurityAnalyticsPlugin extends Plugin implements ActionPlugin, Map
     public static final String FINDINGS_CORRELATE_URI = FINDINGS_BASE_URI + "/correlate";
     public static final String LIST_CORRELATIONS_URI = PLUGINS_BASE_URI + "/correlations";
     public static final String CORRELATION_RULES_BASE_URI = PLUGINS_BASE_URI + "/correlation/rules";
-
+    public static final String TIF_CONFIG_URI = PLUGINS_BASE_URI + "/tif";
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java:114]: threat intel source?

Comment (jowg-amazon) [reply src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java]: updated to `/tif/source`

File: src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java [hunk 2]
```diff
@@ -103,9 +111,12 @@ public class SecurityAnalyticsPlugin extends Plugin implements ActionPlugin, Map
     public static final String FINDINGS_CORRELATE_URI = FINDINGS_BASE_URI + "/correlate";
     public static final String LIST_CORRELATIONS_URI = PLUGINS_BASE_URI + "/correlations";
     public static final String CORRELATION_RULES_BASE_URI = PLUGINS_BASE_URI + "/correlation/rules";
-
+    public static final String TIF_BASE_URI = PLUGINS_BASE_URI + "/tif";
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java:114]: tif_source?

File: src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java [hunk 3]
```diff
@@ -103,9 +111,12 @@ public class SecurityAnalyticsPlugin extends Plugin implements ActionPlugin, Map
...
+    public static final String TIF_SOURCE_CONFIG_URI = PLUGINS_BASE_URI + "/tif/source";
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java:115]: threat_intel/source?

File: src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java [hunk 4]
```diff
@@ -129,6 +140,9 @@ public class SecurityAnalyticsPlugin extends Plugin implements ActionPlugin, Map
     private BuiltinLogTypeLoader builtinLogTypeLoader;
 
     private LogTypeService logTypeService;
+
+    private SATIFSourceConfigDao satifSourceConfigDao;
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java:144]: SaTifSourceConfigDao*

File: src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java [hunk 5]
```diff
@@ -211,13 +228,14 @@ public List<RestHandler> getRestHandlers(Settings settings,
                 new RestSearchCorrelationRuleAction(),
                 new RestIndexCustomLogTypeAction(),
                 new RestSearchCustomLogTypeAction(),
-                new RestDeleteCustomLogTypeAction()
+                new RestDeleteCustomLogTypeAction(),
+                new RestIndexTIFConfigAction()
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java:232]: RestIndexTIFSourceConfigAction

File: src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java [hunk 6]
```diff
@@ -232,7 +250,21 @@ public ScheduledJobRunner getJobRunner() {
...
+        return (xcp, id, jobDocVersion) -> {
+            XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT, xcp.nextToken(), xcp);
+            while (xcp.nextToken() != XContentParser.Token.END_OBJECT) {
+                String fieldName = xcp.currentName();
+                xcp.nextToken();
+                switch (fieldName) {
+                    case FEED_SOURCE_CONFIG_FIELD:
+                        return SATIFSourceConfig.parse(xcp, id, null);
+                    default:
+                        log.warn("Unsupported document was indexed");
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java:262]: log level should be error
re-word message to make it more descriptive

message should include `job parser failed`, `$fieldName` (value), `in security analytics job registration`

File: src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java [hunk 7]
```diff
@@ -232,7 +250,21 @@ public ScheduledJobRunner getJobRunner() {
...
     public ScheduledJobParser getJobParser() {
-        return (parser, id, jobDocVersion) -> TIFJobParameter.PARSER.parse(parser, null);
+        return (xcp, id, jobDocVersion) -> {
+            XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT, xcp.nextToken(), xcp);
+            while (xcp.nextToken() != XContentParser.Token.END_OBJECT) {
+                String fieldName = xcp.currentName();
+                xcp.nextToken();
+                switch (fieldName) {
+                    case FEED_SOURCE_CONFIG_FIELD:
+                        return SATIFSourceConfig.parse(xcp, id, null);
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java:260]: why is the third param null

File: src/main/java/org/opensearch/securityanalytics/threatIntel/action/SAIndexTIFSourceConfigAction.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+ */
+
+package org.opensearch.securityanalytics.threatIntel.action;
+
+import org.opensearch.action.ActionType;
+
+import static org.opensearch.securityanalytics.threatIntel.sacommons.IndexTIFSourceConfigAction.INDEX_TIF_SOURCE_CONFIG_ACTION_NAME;
+
+/**
+ * Threat intel tif job creation action
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/action/SAIndexTIFSourceConfigAction.java:13]: more descriptive and talk about indexing config, downloading feed, setting up job scheduler job

File: src/main/java/org/opensearch/securityanalytics/threatIntel/action/SAIndexTIFSourceConfigRequest.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+import java.time.temporal.ChronoUnit;
+import java.util.List;
+
+/**
+ * Threat intel feed config creation request
+ */
+public class SAIndexTIFSourceConfigRequest extends ActionRequest implements IndexTIFSourceConfigRequest {
+    private static final ParameterValidator VALIDATOR = new ParameterValidator();
+    private String tifSourceConfigId;
+    private final WriteRequest.RefreshPolicy refreshPolicy;
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/action/SAIndexTIFSourceConfigRequest.java:32]: dont accept param. always IMMEDIATE REFRESH

File: src/main/java/org/opensearch/securityanalytics/threatIntel/action/SAIndexTIFSourceConfigResponse.java [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+import java.io.IOException;
+
+import static org.opensearch.securityanalytics.util.RestHandlerUtils._ID;
+import static org.opensearch.securityanalytics.util.RestHandlerUtils._VERSION;
+
+public class SAIndexTIFSourceConfigResponse extends ActionResponse implements ToXContentObject, IndexTIFSourceConfigResponse {
+    private final String tifConfigId;
+    private final Long version;
+    private final RestStatus status;
+    private final SATIFSourceConfigDto saTIFConfigDto;
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/action/SAIndexTIFSourceConfigResponse.java:26]: if a word is an a acronym let's make it liek SaTifSourceConfigDto

File: src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    public SATIFSourceConfigDao(final Client client, final ClusterService clusterService, ThreadPool threadPool) {
+        this.client = client;
+        this.clusterService = clusterService;
+        this.clusterSettings = clusterService.getClusterSettings();
+        this.threadPool = threadPool;
+    }
+
+    public void indexTIFSourceConfig(SATIFSourceConfig satifSourceConfig,
+                                     TimeValue indexTimeout,
+                                     WriteRequest.RefreshPolicy refreshPolicy,
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java:62]: should always be refresh immediate. remove param

File: src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+                                     final ActionListener<SATIFSourceConfig> actionListener) throws Exception {
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java:63]: throws Exception is not needed

File: src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+        client.index(indexRequest, new ActionListener<>() {
+            @Override
+            public void onResponse(IndexResponse response) {
+                log.debug("TIF source config indexed success.");
+                satifSourceConfig.setId(response.getId());
+                actionListener.onResponse(satifSourceConfig);
+            }
+            @Override
+            public void onFailure(Exception e) {
+                throw new SecurityAnalyticsException("Exception saving the tif source config in index", RestStatus.INTERNAL_SERVER_ERROR, e);
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java:78]: use listener.onFailure don't re-throw

File: src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+        this.clusterService = clusterService;
+        this.clusterSettings = clusterService.getClusterSettings();
+        this.threadPool = threadPool;
+    }
+
+    public void indexTIFSourceConfig(SATIFSourceConfig satifSourceConfig,
+                                     TimeValue indexTimeout,
+                                     WriteRequest.RefreshPolicy refreshPolicy,
+                                     final ActionListener<SATIFSourceConfig> actionListener) throws Exception {
+        IndexRequest indexRequest = new IndexRequest(SecurityAnalyticsPlugin.JOB_INDEX_NAME)
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java:64]: wrap entire logic in try-catch and listener.onFailure + log message in catch block

File: src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+                .setRefreshPolicy(refreshPolicy)
+                .source(satifSourceConfig.toXContent(XContentFactory.jsonBuilder(), ToXContent.EMPTY_PARAMS))
+                .timeout(indexTimeout);
+        log.debug("Indexing tif source config");
+        client.index(indexRequest, new ActionListener<>() {
+            @Override
+            public void onResponse(IndexResponse response) {
+                log.debug("TIF source config indexed success.");
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java:72]: we need to log the document id.

File: src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+                satifSourceConfig.setId(response.getId());
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java:73]: copy to new object.

File: src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+    private String getIndexMapping() {
+        try {
+            try (InputStream is = SATIFSourceConfigDao.class.getResourceAsStream("/mappings/threat_intel_job_mapping.json")) {
+                try (BufferedReader reader = new BufferedReader(new InputStreamReader(is, StandardCharsets.UTF_8))) {
+                    return reader.lines().map(String::trim).collect(Collectors.joining());
+                }
+            }
+        } catch (IOException e) {
+            log.error("Runtime exception when getting the threat intel index mapping", e);
+            throw new SecurityAnalyticsException("Runtime exception when getting the threat intel index mapping", RestStatus.INTERNAL_SERVER_ERROR, e);
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java:98]: Message can be "Failed to get threat intel index maping"

File: src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java (hunk 2 above)
Comment (jowg-amazon) [reply src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java]: changed to threat_intel

File: src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java (hunk 3 above)
Comment (jowg-amazon) [reply src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java]: changed to threat_intel/source

File: src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java (hunk 4 above)
Comment (jowg-amazon) [reply src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java]: fixed

File: src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java (hunk 5 above)
Comment (jowg-amazon) [reply src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java]: fixed

File: src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java (hunk 6 above)
Comment (jowg-amazon) [reply src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java]: reworded error message

File: src/main/java/org/opensearch/securityanalytics/threatIntel/action/SAIndexTIFSourceConfigRequest.java (hunk 9 above)
Comment (jowg-amazon) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/action/SAIndexTIFSourceConfigRequest.java:27]: removed param

File: src/main/java/org/opensearch/securityanalytics/threatIntel/action/SAIndexTIFSourceConfigResponse.java (hunk 10 above)
Comment (jowg-amazon) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/action/SAIndexTIFSourceConfigResponse.java]: fixed

File: src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java (hunk 11 above)
Comment (jowg-amazon) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java]: removed

File: src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java (hunk 12 above)
Comment (jowg-amazon) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java]: removed

File: src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java (hunk 14 above)
Comment (jowg-amazon) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java]: done

File: src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+    public void indexTIFSourceConfig(SATIFSourceConfig satifSourceConfig,
+                                     TimeValue indexTimeout,
+                                     WriteRequest.RefreshPolicy refreshPolicy,
+                                     final ActionListener<SATIFSourceConfig> actionListener) throws Exception {
+        IndexRequest indexRequest = new IndexRequest(SecurityAnalyticsPlugin.JOB_INDEX_NAME)
+                .setRefreshPolicy(refreshPolicy)
+                .source(satifSourceConfig.toXContent(XContentFactory.jsonBuilder(), ToXContent.EMPTY_PARAMS))
+                .timeout(indexTimeout);
+        log.debug("Indexing tif source config");
+        client.index(indexRequest, new ActionListener<>() {
```
[the remark this answers is not in the document]
Comment (jowg-amazon) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java]: changed to ActionListener.wrap()

File: src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java (hunk 15 above)
Comment (jowg-amazon) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java]: logged id in message

File: src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java (hunk 13 above)
Comment (jowg-amazon) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java]: done

File: src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java (hunk 17 above)
Comment (jowg-amazon) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java]: Changed message

**Review by eirsep — CHANGES_REQUESTED** [review]

File: src/main/java/org/opensearch/securityanalytics/threatIntel/sacommons/TIFSourceConfigDao.java [hunk 19]
```diff
@@ -0,0 +1,11 @@
+package org.opensearch.securityanalytics.threatIntel.sacommons;
+
+import org.opensearch.core.action.ActionListener;
+public interface TIFSourceConfigDao {
+    IndexTIFSourceConfigResponse indexTIFConfig(IndexTIFSourceConfigRequest request, ActionListener <IndexTIFSourceConfigResponse> listener);
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/sacommons/TIFSourceConfigDao.java:5]: response would be at transport and rest layer?

File: src/main/java/org/opensearch/securityanalytics/threatIntel/service/SATIFSourceConfigService.java [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+                });
+            } catch (Exception e) {
+                listener.onFailure(e);
+            }
+        }, exception -> {
+            lockService.releaseLock(lock);
+            log.error("failed to release lock", exception);
+            listener.onFailure(exception);
+        });
+        SaTifSourceConfigDao.createJobIndexIfNotExists(createIndexStepListener);
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/service/SATIFSourceConfigService.java:79]: this should be done in dao itself before create

File: src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+            }
+
+            @Override
+            public void onFailure(final Exception e) {
+                if (e instanceof ResourceAlreadyExistsException) {
+                    log.info("index[{}] already exist", SecurityAnalyticsPlugin.JOB_INDEX_NAME);
+                    stepListener.onResponse(null);
+                    return;
+                }
+                log.error("Failed to create security analytics threat intel job index", e);
```
[the remark this answers is not in the document]
Comment (jowg-amazon) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java]: Changed to log the index name in error message

File: src/main/java/org/opensearch/securityanalytics/threatIntel/sacommons/TIFSourceConfigDao.java (hunk 19 above)
Comment (jowg-amazon) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/sacommons/TIFSourceConfigDao.java]: Renamed to TIFSourceConfigService and changed to abstract class, will follow up on implementation in future PR.

File: src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java (hunk 7 above)
Comment (jowg-amazon) [reply src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java:270]: Will address this in a future PR, need to understand what params should be used for parser

File: src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java [hunk 22]
```diff
@@ -179,7 +179,7 @@ public Collection<Object> createComponents(Client client,
         TIFJobParameterService tifJobParameterService = new TIFJobParameterService(client, clusterService);
         TIFJobUpdateService tifJobUpdateService = new TIFJobUpdateService(clusterService, tifJobParameterService, threatIntelFeedDataService, builtInTIFMetadataLoader);
         TIFLockService threatIntelLockService = new TIFLockService(clusterService, client);
-        SaTifSourceConfigDao = new SATIFSourceConfigDao(client, clusterService, threadPool);
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/SecurityAnalyticsPlugin.java]: what is this lock service??

File: src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java [hunk 23]
```diff
@@ -48,33 +48,67 @@ public class SATIFSourceConfigDao {
...
+                    log.debug("Threat intel source config with id [{}] indexed success.", response.getId());
+                    SATIFSourceConfig responseSaTifSourceConfig = createSATIFSourceConfig(SaTifSourceConfig, response);
+                    actionListener.onResponse(responseSaTifSourceConfig);
+                }, actionListener::onFailure));
+            } catch (Exception e) {
+                log.error("Exception saving the threat intel source config in index", e);
+                actionListener.onFailure(e);
+            }
+        }, exception -> {
+            lockService.releaseLock(lock);
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/dao/SATIFSourceConfigDao.java:85]: what does this lockservice achieve??

File: src/main/java/org/opensearch/securityanalytics/threatIntel/sacommons/TIFSourceConfigService.java [hunk 24]
```diff
@@ -1,8 +1,10 @@
 package org.opensearch.securityanalytics.threatIntel.sacommons;
 
 import org.opensearch.core.action.ActionListener;
-public interface TIFSourceConfigDao {
-    IndexTIFSourceConfigResponse indexTIFConfig(IndexTIFSourceConfigRequest request, ActionListener <IndexTIFSourceConfigResponse> listener);
+public abstract class TIFSourceConfigService {
+    IndexTIFSourceConfigResponse indexTIFConfig(IndexTIFSourceConfigRequest request, ActionListener <IndexTIFSourceConfigResponse> listener){
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/sacommons/TIFSourceConfigService.java:5]: method should be abstract

**Review by eirsep — APPROVED** [review]
plz address comments in followup PR

**Pull request merged**